internal/grpcsync/pubsub_test.go

```diff
@@ -0,0 +1,211 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpcsync
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type testSubscriber struct {
+	mu      sync.Mutex
+	msgs    []int
+	onMsgCh chan struct{}
+}
+
+func newTestSubscriber(chSize int) *testSubscriber {
+	return &testSubscriber{onMsgCh: make(chan struct{}, chSize)}
+}
+
+func (ts *testSubscriber) OnMessage(msg interface{}) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	ts.msgs = append(ts.msgs, msg.(int))
+	select {
+	case ts.onMsgCh <- struct{}{}:
+	default:
+	}
+}
+
+func (ts *testSubscriber) receivedMsgs() []int {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	msgs := make([]int, len(ts.msgs))
+	copy(msgs, ts.msgs)
+
+	return msgs
+}
+
+func (s) TestPubSub_PublishNoMsg(t *testing.T) {
+	pubsub := NewPubSub()
+	defer pubsub.Stop()
+
+	ts := newTestSubscriber(1)
+	pubsub.Subscribe(ts)
+
+	select {
+	case <-ts.onMsgCh:
+		t.Fatalf("Subscriber callback invoked when no message was published")
+	case <-time.After(defaultTestShortTimeout):
+	}
+}
+
+func (s) TestPubSub_PublishMsgs_RegisterSubs_And_Stop(t *testing.T) {
+	pubsub := NewPubSub()
+
+	const numPublished = 10
+
+	ts1 := newTestSubscriber(numPublished)
+	pubsub.Subscribe(ts1)
+	wantMsgs1 := []int{}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	// Publish ten messages on the pubsub and ensure that they are received in order by the subscriber.
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			pubsub.Publish(i)
+			wantMsgs1 = append(wantMsgs1, i)
+		}
+		wg.Done()
+	}()
+
+	isTimeout := false
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			select {
+			case <-ts1.onMsgCh:
```

Comment:
IIUC, we are trying to check the subscriber receieves the messages in the order they are published. And we know the message that is published in each iteration (which is `i := range numPublished`). Why dont we just check for the message received here and fail if its not the one that is expected? 

that way we can nix `wantMsg1` and `gotMsg1` variables and also fail faster. 

I usually like the idea of always failing fast as possible in test assertions. Here lets say that the first message itself came out of order, we would have to wait until all the messages are received to fail the test.

Reply:
In this case, the one sub goroutine is used to call `Publish()` and the other is used to check to see if `ts1.OnMessage()` function was called the correct number of times.
Both are processed asynchronously.

Also, if we tried to fail faster, it would be more complicated than before.
(I haven't come up with the idea to do yet.)